flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java

```diff
@@ -99,45 +100,90 @@ private static BlobStoreService createFileSystemBlobStore(Configuration configur
+            @Nullable InetSocketAddress serverAddress)
+            throws IOException {
+        final Reference<File> storageDirectory =
+                createBlobStorageDirectory(configuration, fallbackStorageDirectory);
+        return new BlobCacheService(configuration, storageDirectory, blobView, serverAddress);
+    }
+
+    static Reference<File> createBlobStorageDirectory(
+            Configuration configuration, @Nullable Reference<File> fallbackStorageDirectory)
+            throws IOException {
+        final String basePath = configuration.getString(BlobServerOptions.STORAGE_DIRECTORY);
 
-        File baseDir;
+        File baseDir = null;
         if (StringUtils.isNullOrWhitespaceOnly(basePath)) {
-            final String[] tmpDirPaths = ConfigurationUtils.parseTempDirectories(config);
-            baseDir = new File(tmpDirPaths[RANDOM.nextInt(tmpDirPaths.length)]);
+            if (fallbackStorageDirectory != null) {
+                baseDir = fallbackStorageDirectory.deref();
+
+                if (baseDir.mkdirs() || baseDir.exists()) {
+                    return fallbackStorageDirectory;
+                }
+            }
         } else {
             baseDir = new File(basePath);
-        }
 
-        File storageDir;
+            File storageDir;
 
-        // NOTE: although we will be using UUIDs, there may be collisions
-        int maxAttempts = 10;
-        for (int attempt = 0; attempt < maxAttempts; attempt++) {
-            storageDir =
-                    new File(baseDir, String.format("blobStore-%s", UUID.randomUUID().toString()));
+            // NOTE: although we will be using UUIDs, there may be collisions
+            int maxAttempts = 10;
+            for (int attempt = 0; attempt < maxAttempts; attempt++) {
+                storageDir =
+                        new File(
+                                baseDir,
+                                String.format("blobStore-%s", UUID.randomUUID().toString()));
 
-            // Create the storage dir if it doesn't exist. Only return it when the operation was
-            // successful.
-            if (storageDir.mkdirs()) {
-                return storageDir;
+                // Create the storage dir if it doesn't exist. Only return it when the operation was
+                // successful.
+                if (storageDir.mkdirs()) {
+                    return Reference.owned(storageDir);
+                }
             }
         }
 
-        // max attempts exceeded to find a storage directory
-        throw new IOException(
-                "Could not create storage directory for BLOB store in '" + baseDir + "'.");
+        if (baseDir != null) {
+            throw new IOException(
+                    "Could not create storage directory for BLOB store in '" + baseDir + "'.");
+        } else {
+            throw new IOException(
+                    "Could not create storage directory for BLOB store because not storage directory has been specified.");
```

Comment:
```suggestion
                    "Could not create storage directory for BLOB store because no storage directory has been specified.");
```

Reply:
maybe also include the corresponding config key

Reply:
Will do.